[System.Native] check `ro.build.version.sdk_full` for minor Android releases

jonathanpeppers opened the pull request:
Context: https://developer.android.com/about/versions/16/qpr2/
Context: https://github.com/dotnet/android/commit/35d471e71c986c302dbf019e5c7838e57d3e0a3e
Context: https://developer.android.com/reference/android/os/Build.VERSION#SDK_INT_FULL

Android 16 Quarterly Platform Release 2 (QPR2) Beta 2 has been released.

Android 16 QPR2 adds a new concept, "minor" SDK versions:

    > The new [`SDK_INT_FULL`][0] constant can be used for API checks…
    >
    >     if (SDK_INT_FULL >= VERSION_CODES_FULL.[MAJOR or MINOR RELEASE]) {
    >       // Use APIs introduced in a major or minor release
    >     }
    >
    > You can also use the [`Build.getMinorSdkVersion()`][1] method to
    > get just the minor SDK version:
    >
    >     minorSdkVersion = Build.getMinorSdkVersion(Build.VERSION_CODES_FULL.BAKLAVA);

To mirror this in .NET, we introduced a new `$(TargetFramework)` such as `net10.0-android36.1` where C# developers could use:

    if (OperatingSystem.IsAndroidVersionAtLeast(36, 1))
    {
        // Use APIs introduced in Android 36.1
    }

Reviewing Android's implementation of `Build.VERSION.SDK_INT_FULL` we can check the `ro.build.version.sdk_full` system property to get the minor version and falling back to `ro.build.version.sdk` otherwise.

`libSystem.Native.so` can return the value from `SystemNative_GetUnixRelease()` making both `Environment.OSVersion` and `System.IsAndroidVersionAtLeast()` support minor Android releases.

An example of the system properties available on the latest emulator:

    % adb shell getprop | grep ro.build.version.sdk
    [ro.build.version.sdk]: [36]
    [ro.build.version.sdk_full]: [36.1]

The release notes mention QPR2 Beta 2 is API stable, so we should be safe to rely on this behavior going forward:

> QPR2 Beta 2 is now available and has released Platform Stability.
> That means that the API surface is locked, and the app-facing behaviors
> are final, so you can incorporate them into your apps and take advantage
> of our latest platform innovations.

To test this behavior, I set `$DOTNET_ANDROID_VERSION_FROM_JAVA` in the Android test runner, to verify we get the same value from Java and .NET.

[0]: https://developer.android.com/reference/android/os/Build.VERSION#SDK_INT_FULL
[1]: https://developer.android.com/reference/android/os/Build#getMinorSdkVersion(int)

jonathanpeppers committed:
[System.Native] check `ro.build.version.sdk_full` for minor Android releases

Context: https://developer.android.com/about/versions/16/qpr2/
Context: https://github.com/dotnet/android/commit/35d471e71c986c302dbf019e5c7838e57d3e0a3e
Context: https://developer.android.com/reference/android/os/Build.VERSION#SDK_INT_FULL

Android 16 Quarterly Platform Release 2 (QPR2) Beta 2 has been released.

Android 16 QPR2 adds a new concept, "minor" SDK versions:

    > The new [`SDK_INT_FULL`][0] constant can be used for API checks…
    >
    >     if (SDK_INT_FULL >= VERSION_CODES_FULL.[MAJOR or MINOR RELEASE]) {
    >       // Use APIs introduced in a major or minor release
    >     }
    >
    > You can also use the [`Build.getMinorSdkVersion()`][1] method to
    > get just the minor SDK version:
    >
    >     minorSdkVersion = Build.getMinorSdkVersion(Build.VERSION_CODES_FULL.BAKLAVA);

To mirror this in .NET, we introduced a new `$(TargetFramework)` such as
`net10.0-android36.1` where C# developers could use:

    if (OperatingSystem.IsAndroidVersionAtLeast(36, 1))
    {
        // Use APIs introduced in Android 36.1
    }

Reviewing Android's implementation of `Build.VERSION.SDK_INT_FULL` we
can check the `ro.build.version.sdk_full` system property to get the
minor version and falling back to `ro.build.version.sdk` otherwise.

`libSystem.Native.so` can return the value from
`SystemNative_GetUnixRelease()` making both `Environment.OSVersion`
and `System.IsAndroidVersionAtLeast()` support minor Android releases.

An example of the system properties available on the latest emulator:

    % adb shell getprop | grep ro.build.version.sdk
    [ro.build.version.sdk]: [36]
    [ro.build.version.sdk_full]: [36.1]

The release notes mention QPR2 Beta 2 is API stable, so we should be
safe to rely on this behavior going forward:

> QPR2 Beta 2 is now available and has released Platform Stability.
> That means that the API surface is locked, and the app-facing behaviors
> are final, so you can incorporate them into your apps and take advantage
> of our latest platform innovations.

To test this behavior, I set `$DOTNET_ANDROID_VERSION_FROM_JAVA` in the
Android test runner, to verify we get the same value from Java and .NET.

[0]: https://developer.android.com/reference/android/os/Build.VERSION#SDK_INT_FULL
[1]: https://developer.android.com/reference/android/os/Build#getMinorSdkVersion(int)

jonathanpeppers commented on src/tasks/AndroidAppBuilder/Templates/MonoRunner.java:
```diff
@@ -188,6 +189,20 @@ static int getLocalDateTimeOffset() {
+        
+        // Check if Build.getMinorSdkVersion() is available (Android 16+ / API 36+)
+        if (android.os.Build.VERSION.SDK_INT >= 36) {
+            int minorSdk = android.os.Build.getMinorSdkVersion(android.os.Build.VERSION.SDK_INT_FULL);
+            if (minorSdk > 0) {
+                return majorSdk + "." + minorSdk;
```
They _are_ available on 36+:
* https://developer.android.com/reference/android/os/Build.VERSION#SDK_INT_FULL
* https://developer.android.com/reference/android/os/Build#getMinorSdkVersion(int)

akoeplinger replied:
I'm not sure we always have a new enough Android SDK that contains SDK_INT_FULL

jonathanpeppers committed:
Remove test changes

These were useful to at least test the new code works:

09-17 16:41:13.701  7106  7121 D DOTNET  : Setting env var: DOTNET_ANDROID_VERSION_FROM_JAVA=36.1
...
09-17 16:41:14.927  7106  7130 I DOTNET  :     [PASS] System.Tests.EnvironmentTests.OSVersion_ValidVersion_AndroidMinor

akoeplinger commented:
/backport to release/10.0
